### PR TITLE
Added movement handles to 2D

### PR DESCRIPTION
**Pull request by MarcusElg**
There are already nice looking handles for scaling so why not for movement? This makes it easier to move an object on a single axis and also increases consistency between 3d/2d as 3d already has movement handles.
Movement Handles

It's a small neat quality of life feature that makes the 2d editor feel more robust. Let me know what you think!

EDIT: As requested I added a way to disable the handle (both visually and it's function):

### CODE REVIEW

**Comment by KoBeWi** [comment]
> This makes it easier to move an object on a single axis

Shift already does that.

**Comment by asheraryam** [comment]
@KoBeWi So does G in blender but they should the widget by default 🤷‍♂ 

As long as it's possible to disable widgets in 2D like the current options in 3D viewport it should be fine.

**Comment by MCrafterzz** [comment]
> @KoBeWi Using shift is much faster but so is G in blender but they show the widget by default 
> 
> As long as it's possible to disable widgets in 2D like the current options in 3D viewport it should be fine.

Adding a iption to disable widgets is a good idea but I feel like that should be in a separate pr

**Comment by KoBeWi** [comment]
>  I feel like that should be in a separate pr

If you plan to do this anyways, you could make it as second commit in this PR.

**Comment by MCrafterzz** [comment]
@asheraryam I've looked around and haven't found any way to disable it in 3D. So can you explain how to do that so that I can implement similar behaviour for 2D?

**Comment by NathanLovato** [comment]
Having visual widgets is important, for two reasons:

- Discoverability. You shouldn't assume every user understands the possible move interactions.
- Clear visual feedback on the active tool. It is clear that in this case, you have the move tool active, and not rotate or scale. Right now, as a tutor, I often have to point people at the icons in the toolbar as it is not obvious which tool is currently active.

As far as letting people toggle the widget inactive, I think that the solution the Blender team implemented is excellent. You have a single icon to toggle all gizmos visible in the 3d viewport.
On:

Screenshot from 2019-11-09 21-11-51

Off:

Screenshot from 2019-11-09 21-12-01

A sub-menu allows you to customize the gizmos you want to display, e.g. move, scale, rotate, but also their mouse-based view navigation widget.

**Comment by Calinou** [comment]
@MCrafterzz Hiding all gizmos can be done in the menu at the top-left of the 3D viewport:

Perspective menu

The **View** menu can also be used to toggle individual gizmos:

View menu

**Comment by MCrafterzz** [comment]
@Calinou Even though show gizmos is disabled and all gizmos in the dialog are as well they still show movement/rotation/scale handles but I guess I could add it to 2D anyways. Also do I have to add options for all those other stuff (joints etc) or can that come later?

**Comment by MCrafterzz** [comment]
@Calinou @NathanLovato @asheraryam @KoBeWi I finally managed to add a menu option for disabling the movement and scaling gizmo. See the original message for a pictures. It's clean and easy to find, so it it enough?

**Comment by Calinou** [comment]
@MCrafterzz Is there an use for toggling each gizmo individually, or could we put it behind a global toggle? This way, we wouldn't need an extra dropdown menu.

**Comment by MCrafterzz** [comment]
> @MCrafterzz Is there an use for toggling each gizmo individually, or could we put it behind a global toggle? This way, we wouldn't need an extra dropdown menu.

@Calinou 
Yes. There is no good way to scale one axis(as far as I know) so most people would want to use the scale handles but some oppose this new movement handle

**Comment by groud** [comment]
I think the feature is ok, but as stated above, a setting per gizmo is pointless. A single setting is ok for both.

That being said, I am not sure it make sense to hide the gizmo when the scale tool is active. What would be the point of this mode otherwise ?

**Comment by KoBeWi** [comment]
> That being said, I am not sure it make sense to hide the gizmo when the scale tool is active. What would be the point of this mode otherwise ?

You don't need to use the gizmo to scale, but it's difficult. Disabling it is like activating Godot hard mode.

**Comment by MCrafterzz** [comment]
Ok I'll merge it into one option tommorow

**Comment by MCrafterzz** [comment]
@groud @KoBeWi Merged it into one option:

**Comment by aaronfranke** [comment]
@MCrafterzz I'm surprised nobody has pointed this out yet... but the green arrow needs to point down.

**Comment by MCrafterzz** [comment]
> @MCrafterzz Surprised nobody has pointed this out yet... but the green arrow needs to point down.

@aaronfranke Should I do the same for the scale handle? It's kinda strange if they point in two different ways. Also WHY is positive downwards? Is there another program in the world that does that? It feels very strange an unlogical :P

**Comment by groud** [comment]
> @aaronfranke Should I do the same for the scale handle?

I guess it makes sense.

> Also WHY is positive downwards? Is there another program in the world that does that? 

Almost all other software related to 2D editing. Numbering and positioning does almost all the time use the top left as the origin point, this make sense as this is consistent with the reading direction in English.

**Comment by MCrafterzz** [comment]
> @aaronfranke Should I do the same for the scale handle?
> 
> I guess it makes sense.
> 
> Also WHY is positive downwards? Is there another program in the world that does that?
> 
> Almost all other software related to 2D editing. Numbering and positioning does almost all the time use the top left as the origin point, this make sense as this is consistent with the reading direction in English.

Huh, I had no idea that other programs had it that way. I guess it's a logical explanation but it still looks very strange to have the arrow point downwards :P

**Comment by MCrafterzz** [comment]
@groud I've now changed it. Is there anything more I need to do or is this finished? The three comment do different stuff but should I squash them together anyways?

**Comment by MCrafterzz** [comment]
@KoBeWi I want to merge the last three commits together but I can't use git rebase as I usually do as it lists all the master commits (due to me fixing the merge conflicts). What should I do?

**Comment by KoBeWi** [comment]
No idea tbh, I had it happen once and managed to fix it by doing usual rebase (I think the merge ended up as one commit in the end, or maybe I just did a huge fixup on tens of commits). It then happened again in another PR and I messed it so much that I just closed it and make a new one :I

I usually just avoid running into this kind of problems. If you haven't pushed it yet, you could try reverting your local changes and merging the commits before resolving conflicts.

**Comment by aaronfranke** [comment]
@MCrafterzz Never push a "Merge branch 'master' into" commit, it's bad history.

The conflicts should be fine to resolve in favor of your branch. With your branch checked out, run these commands:

    git reset --hard HEAD~2
    git clean -fd
    git rebase -Xtheirs master
    # you should build and test everything at this point to make sure it works
    git push --force

The first command removes the "Merge branch 'master' into" commits, the second cleans up (there's new things in the gitignore which weren't there before, so this is indeed necessary), the third command rebases resolving in favor of your copy, and the last command force pushes. The rebase can also be done with a simple drag-and-drop in [my favorite Git client](https://www.youtube.com/watch?v=0XKy8YpP5Ww).

**Comment by MCrafterzz** [comment]
@aaronfranke Thanks that solved it! What is the best way to prevent this issue in the future?

**Comment by MCrafterzz** [comment]
@akien-mga I'm finally happy to say that this pr is finished to be merge 3 months later. Hopefullt there won't be another conflict before this is merged :P

**Comment by MCrafterzz** [comment]
@akien-mga Or is it to early to merge new features?

**Comment by Calinou** [comment]
@MCrafterzz Most Godot core developers are currently busy at GodotCon :slightly_smiling_face:

**Comment by MCrafterzz** [comment]
@Calinou you can merge it :D

**Comment by groud** [comment]
No, I need to review this first (on a computer). It will have to wait a little bit :)

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 1]
```diff
@@ -3211,10 +3240,49 @@ void CanvasItemEditor::_draw_selection() {
...
+
+					Size2 move_factor = Size2(MOVE_HANDLE_DISTANCE, MOVE_HANDLE_DISTANCE);
+					viewport->draw_set_transform_matrix(simple_xform);
+
+					Vector<Point2> points;
+					points.push_back(Vector2(move_factor.x * EDSCALE, 5 * EDSCALE));
+					points.push_back(Vector2(move_factor.x * EDSCALE, -5 * EDSCALE));
+					points.push_back(Vector2((move_factor.x + 10) * EDSCALE, 0));
+
+					Vector<Color> colors;
```
Comment (groud) [inline editor/plugins/canvas_item_editor_plugin.cpp]: Couldn't you use `draw_colored_polygon` instead of pushing 3 colors every time ?

Comment (MCrafterzz) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Implemeted. Yes that's defiantly the better way

**Comment by MCrafterzz** [comment]
@groud anything more?

**Comment by groud** [comment]
This PR LGTM at its current state. :) @akien-mga might want you to squash the commits together though.

**Comment by MCrafterzz** [comment]
> This PR LGTM at its current state. :) @akien-mga might want you to squash the commits together though.

It was ment to be 3 prs but people thought it would better to do it in the same and have it as separate commits. Every commit does something different.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**